### ISSUE TITLE
REST API /V1/products/attributes does not update all properties

### ISSUE THREAD

**Issue by sazzle2611** [magento/magento2#10205]
Using Magento 2.1.7 with the REST API - POST /V1/products/attributes - I can add a new attribute and configure it with settings such as is_filterable.

This is all fine but if I then want to change this setting I can't. I have tried with POST and with PUT /V1/products/attributes/{attributeCode}. The command works as in it will update, for instance, the frontend label in the eav_attribute table, but it does not seem to have any effect on values in the catalog_eav_attribute table.

Steps to reproduce -
1. Add a new attribute using REST API - POST - /V1/products/attributes - with is_filterable set to false.
2. Update the same attribute using either the same POST url or PUT - /V1/products/attributes/{attributeCode} and set the is_filterable field to true

Expected result
The attribute is now set to be filterable

Actual result
No change

**Comment by misha-kotov** [comment]
Hi @sazzle2611 thanks for reporting this issue. 
Do you mind adding some more details in accordance to our [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines): with steps to reproduce, actual result and expected result. It will help us reproduce and troubleshoot the issue.

**Comment by magento-engcom-team** [comment]
@sazzle2611 we are closing this issue due to inactivity. If you'd like to update it, please reopen the issue.

**Comment by tkotosz** [comment]
I can reproduce the same issue in magento 2.2.2. I have tried to update the visibility of an attribute (`is_visible_on_front`) with POST request to `/V1/products/attributes` and PUT request to `/V1/products/attributes/{attribute_code}` but it didn't change the attribute property.

Update: it seems if I do a PUT `/V1/products/attributes/{attribute_code}` and provide the attribute_id and the entity type id as well then I can change the visibility like this:
```
{
  "attribute": {
        "is_visible_on_front": "0",
        "attribute_id": 4603,
        "entity_type_id": "4"
    }
}
```
which is really weird. (note that without the entity_type_id it doesn't work)

**Comment by engcom-backlog-nickolas** [comment]
Hello @sazzle2611, thank you for report. Unfortunately I wasn't able to reproduce this issue.
![peek 2018-08-10 10-59](https://user-images.githubusercontent.com/40755810/43945728-bd40322c-9c8b-11e8-99bb-56c953cbb720.gif)
If you'd like to update it, please reopen the issue.

@tkotosz it's correct behaviour. You have to specify entity type id, because is_visible_on_front property related to product entity.

**Comment by lewisvoncken** [comment]
@engcom-backlog-nickolas Yeah I understand but still the call is already specified for product attributes based on the API call so why do you still need to add the entity type id to the body request?

**Comment by sidolov** [comment]
Hi @sazzle2611. Thank you for your report. 
The issue has been fixed in magento/magento2#18622 by @lewisvoncken in 2.3-develop branch
Related commit(s):
 - [42ecb9c78de2ebb4c1c475b2ef2e286faded0b5f](https://github.com/magento/magento2/commit/42ecb9c78de2ebb4c1c475b2ef2e286faded0b5f)

The fix will be available with the upcoming 2.3.1 release.

**Comment by ghost** [comment]
Hi @sazzle2611 , thank you for you report, this issue has already fixed in 2.3-develop branch, and will be available on 2.3.1 release.

**Comment by carloricci84** [comment]
On 2.2.3 i made a plugin like this to fix the issue since we can't upgrade to 2.3 yet :

Vendor\Module\di.xml
```
<type name="Magento\Catalog\Model\Product\Attribute\Repository">
    <plugin name="vendor_attribute_repository_plugin" type="Vendor\Module\Plugin\AttributeRepositoryPlugin" />
</type>
```

Vendor\ModulePlugin\AttributeRepositoryPlugin.php

```
namespace Vendor\Module\Plugin;

class AttributeRepositoryPlugin
{

    public function __construct(
        \Magento\Eav\Model\Config $eavConfig
    ) {
        $this->eavConfig = $eavConfig;
    }

    public function beforeSave (\Magento\Catalog\Model\Product\Attribute\Repository $subject, \Magento\Catalog\Api\Data\ProductAttributeInterface $attribute ) {
        $attribute->setEntityTypeId(
            $this->eavConfig
                ->getEntityType(\Magento\Catalog\Api\Data\ProductAttributeInterface::ENTITY_TYPE_CODE)
                ->getId()
        );

        return [$attribute];
    }
}
```
